### PR TITLE
Utils library and some modernization

### PR DESCRIPTION
**Pull request by fweik**
Sorry that got a bit out of hand:
- removed `OLD_DIHEDRAL` feature
- removed `sqrlen`, `scalar`, `vector_product` for c-arrays
- new `Utils::gaussian` and `Utils::raster` functions
- replaced various c-style arrays by Utils::Vector
- modernization and const correctness for interface of pair bonds
- moved utils to separate library to help with code restructuring
- removed `utils.hpp`

### CODE REVIEW

File: src/core/accumulators/Correlator.cpp [hunk 1]
```diff
@@ -20,7 +20,6 @@
 #include "integrate.hpp"
 #include "partCfg_global.hpp"
 #include "particle_data.hpp"
-#include "utils.hpp"
 
 #include "utils/serialization/multi_array.hpp"
```
Comment (KaiSzuttor) [inline src/core/accumulators/Correlator.cpp]: ```<``` [...] ```>```

Comment (KaiSzuttor) [reply src/core/accumulators/Correlator.cpp]: should be corrected in all places

File: src/core/bonded_interactions/bonded_coulomb.hpp [hunk 2]
```diff
@@ -55,29 +54,18 @@ int bonded_coulomb_set_params(int bond_type, double prefactor);
...
                                           Particle const *p2,
                                           Bonded_ia_parameters const *iaparams,
-                                          double const dx[3], double force[3]) {
+                                          const Utils::Vector3d &dx,
+                                          double *force) {
   int i;
   double fac;
-  double dist2 = sqrlen(dx);
-  double dist = sqrt(dist2);
+  const double dist2 = dx.norm2();
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb.hpp]: east const

File: src/core/bonded_interactions/bonded_coulomb.hpp [hunk 3]
```diff
@@ -55,29 +54,18 @@ int bonded_coulomb_set_params(int bond_type, double prefactor);
 inline int calc_bonded_coulomb_pair_force(Particle const *p1,
                                           Particle const *p2,
                                           Bonded_ia_parameters const *iaparams,
-                                          double const dx[3], double force[3]) {
+                                          const Utils::Vector3d &dx,
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb.hpp]: east const

File: src/core/bonded_interactions/bonded_coulomb.hpp [hunk 4]
```diff
@@ -92,8 +80,9 @@ inline int calc_bonded_coulomb_pair_force(Particle const *p1,
  */
 inline int bonded_coulomb_pair_energy(Particle const *p1, Particle const *p2,
                                       Bonded_ia_parameters const *iaparams,
-                                      double const dx[3], double *_energy) {
-  double dist = sqrt(sqrlen(dx));
+                                      const Utils::Vector3d &dx,
+                                      double *_energy) {
+  double dist = dx.norm();
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb.hpp]: const

File: src/core/bonded_interactions/bonded_coulomb_sr.hpp [hunk 5]
```diff
@@ -54,12 +53,11 @@ int bonded_coulomb_sr_set_params(int bond_type, double q1q2);
  */
 inline int
 calc_bonded_coulomb_sr_pair_force(Bonded_ia_parameters const *iaparams,
-                                  double dx[3], double force[3]) {
-  double dist2 = sqrlen(dx);
-  double dist = sqrt(dist2);
+                                  const Utils::Vector3d &dx, double *force) {
+  double dist = dx.norm();
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb_sr.hpp]: const

File: src/core/bonded_interactions/bonded_coulomb_sr.hpp [hunk 6]
```diff
@@ -78,11 +76,13 @@ calc_bonded_coulomb_sr_pair_force(Bonded_ia_parameters const *iaparams,
  */
 inline int bonded_coulomb_sr_pair_energy(const Particle *p1, const Particle *p2,
                                          Bonded_ia_parameters const *iaparams,
-                                         double *dx, double *_energy) {
-  double dist2 = sqrlen(dx);
+                                         const Utils::Vector3d &dx,
+                                         double *_energy) {
+  double dist2 = dx.norm2();
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/bonded_coulomb_sr.hpp]: const

File: src/core/bonded_interactions/harmonic.hpp [hunk 7]
```diff
@@ -40,19 +41,15 @@
...
  *  @retval 1 if the bond is broken
  *  @retval 0 otherwise
  */
-inline int calc_harmonic_pair_force(Particle const *p1, Particle const *p2,
-                                    Bonded_ia_parameters const *iaparams,
-                                    double const dx[3], double force[3]) {
-  double dist2 = sqrlen(dx);
-  double dist = sqrt(dist2);
+inline int calc_harmonic_pair_force(Bonded_ia_parameters const *iaparams,
+                                    const Utils::Vector3d &dx, double *force) {
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/harmonic.hpp]: east const

File: src/core/bonded_interactions/harmonic.hpp [hunk 8]
```diff
@@ -40,19 +41,15 @@
...
+  double dist = dx.norm();
```
Comment (KaiSzuttor) [inline src/core/bonded_interactions/harmonic.hpp]: const

File: src/core/immersed_boundary/ibm_tribend.cpp [hunk 9]
```diff
@@ -40,45 +40,33 @@ void IBM_Tribend_CalcForce(Particle *p1, Particle *p2, Particle *p3,
...
     sc = 1.0;
 
   // Get theta as angle between normals
   double theta = acos(sc);
 
-  double direc[3];
-  vector_product(n1, n2, direc);
-  const double desc = scalar(dx1, direc);
+  auto const direc = vector_product(n1, n2);
+  const double desc = (dx1 * direc);
```
Comment (KaiSzuttor) [inline src/core/immersed_boundary/ibm_tribend.cpp:69]: Death to the traitor! :point_up_2: :point_up: :point_up_2: :point_up:

File: src/core/immersed_boundary/ibm_triel.cpp [hunk 10]
```diff
@@ -321,20 +322,17 @@ int IBM_Triel_SetParams(const int bond_type, const int ind1, const int ind2,
 
   // Calculate equilibrium lengths and angle; Note the sequence of the points!
   // lo = length between 1 and 3
-  double templo[3];
-  get_mi_vector(templo, part3.r.p, part1.r.p);
-  const double l0 = sqrt(sqrlen(templo));
+  auto const templo = get_mi_vector(part3.r.p, part1.r.p);
+  const double l0 = templo.norm();
```
Comment (KaiSzuttor) [inline src/core/immersed_boundary/ibm_triel.cpp:326]: oh no

File: src/core/immersed_boundary/ibm_triel.cpp [hunk 11]
```diff
@@ -321,20 +322,17 @@ int IBM_Triel_SetParams(const int bond_type, const int ind1, const int ind2,
...
   // lpo = length between 1 and 2
-  double templpo[3];
-  get_mi_vector(templpo, part2.r.p, part1.r.p);
-  const double lp0 = sqrt(sqrlen(templpo));
+  auto const templpo = get_mi_vector(part2.r.p, part1.r.p);
+  const double lp0 = templpo.norm();
```
Comment (KaiSzuttor) [inline src/core/immersed_boundary/ibm_triel.cpp:329]: oh no

File: src/core/immersed_boundary/ibm_triel.cpp [hunk 12]
```diff
@@ -321,20 +322,17 @@ int IBM_Triel_SetParams(const int bond_type, const int ind1, const int ind2,
...
 
   // cospo / sinpo angle functions between these vectors; calculated directly
   // via the products
-  const double cosPhi0 = scalar(templo, templpo) / (l0 * lp0);
-  double vecpro[3];
-  vector_product(templo, templpo, vecpro);
-  const double sinPhi0 = sqrt(sqrlen(vecpro)) / (l0 * lp0);
+  const double cosPhi0 = (templo * templpo) / (l0 * lp0);
```
Comment (KaiSzuttor) [inline src/core/immersed_boundary/ibm_triel.cpp:333]: oh no

File: src/core/immersed_boundary/ibm_triel.cpp [hunk 13]
```diff
@@ -321,20 +322,17 @@ int IBM_Triel_SetParams(const int bond_type, const int ind1, const int ind2,
...
+  auto const vecpro = vector_product(templo, templpo);
+  const double sinPhi0 = vecpro.norm() / (l0 * lp0);
```
Comment (KaiSzuttor) [inline src/core/immersed_boundary/ibm_triel.cpp:335]: oh no

File: src/core/pair_criteria/pair_criteria.hpp [hunk 14]
```diff
@@ -63,15 +61,14 @@ class EnergyCriterion : public PairCriterion {
 public:
   bool decide(const Particle &p1, const Particle &p2) const override {
     // Distance between particles
-    double vec21[3];
-    get_mi_vector(vec21, p1.r.p, p2.r.p);
-    const double dist_betw_part = sqrt(sqrlen(vec21));
+    auto const vec21 = get_mi_vector(p1.r.p, p2.r.p);
+    const double dist_betw_part = vec21.norm();
```
Comment (KaiSzuttor) [inline src/core/pair_criteria/pair_criteria.hpp:65]: oh no

**Review by KaiSzuttor — APPROVED** [review]

File: src/core/immersed_boundary/ibm_tribend.cpp (hunk 9 above)
Comment (fweik) [reply src/core/immersed_boundary/ibm_tribend.cpp:69]: Didn't we agree not to enforce this because there is no tool to check it? ^^

Comment (KaiSzuttor) [reply src/core/immersed_boundary/ibm_tribend.cpp:69]: but it makes me happy :smiling_imp: that's why I approved nonetheless

**Comment by RudolfWeeber** [comment]
On Fri, Apr 26, 2019 at 05:36:53AM -0700, Kai Szuttor wrote:
> +  const double l0 = templo.norm();
> 
> oh no
I'd say let's not bother about the position of const for boring data types,
unless there is a tool which does it automatically.

**Comment by KaiSzuttor** [comment]
@RudolfWeeber my comments were not meant to be taken too serious... That's why I approved. Sorry for the noise.

**Pull request merged**